yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnExtensionService.scala

```diff
@@ -0,0 +1,198 @@
+ *
+ * An instance may be configured to register a shutdown hook. This addresses the issue in
+ * which a spark context is not always closed in an application, especially a `yarn-client`
+ * application. The shutdown hook allows services an opportunity to perform any final
+ * actions their `Service.stop` invocations trigger.
+ */
+private[spark] class YarnExtensionServices extends AbstractService("YarnExtensionServices")
+    with Logging {
+  private var services: List[YarnExtensionService] = Nil
+  private var sparkContext: SparkContext = _
+  private var appId: ApplicationId = _
+  private var attemptId: Option[ApplicationAttemptId] = _
+  private val started = new AtomicBoolean(false)
+  private var registerShutdownHook = false
+  private var shutdownHookPriority = 0
+
+  /**
+   * This is the routine to optionally be registered as a shutdown hook.
+   * It logs the invocation and then calls the `close()` method to stop the
+   * service instance. It is public to aid in testing.
+   */
+  val shutdownHook = new Runnable {
+    override def run(): Unit = {
+      logInfo(s"In shutdown hook for $this")
+      try {
+        stop()
+      }
+      catch {
+        case e: Exception =>
+          logInfo(s"During shutdown: $e", e)
+      }
+    }
+  }
+
+  /**
+   * Binding operation will load the named services and call bind on them too; the
+   * entire set of services are then ready for `init()` and `start()` calls
+   * @param context spark context
+   * @param appId application ID
+   * @param attemptId YARN attempt ID if known. Otherwise, `None`.
+   * @param registerShutdownHook should the services register as a hadoop shutdown hook?
+   * @param shutdownHookPriority and if so, what priority
+   */
+  def start(context: SparkContext,
+      appId: ApplicationId,
+      attemptId: Option[ApplicationAttemptId] = None,
+      registerShutdownHook: Boolean = false,
+      shutdownHookPriority: Int = 0): Unit = {
+    if (isInState(STATE.STARTED)) {
+      logWarning("Ignoring re-entrant start operation")
+      return
+    }
+    require(context != null, "Null context parameter")
+    require(appId != null, "Null appId parameter")
+    require(!registerShutdownHook || shutdownHookPriority > 0,
+           "if registerShutdownHook is set, shutdownHookPriority must be greater than zero")
+    // this operation will trigger the state change checks
+    super.init(context.hadoopConfiguration)
+    this.sparkContext = context
+    this.appId = appId
+    this.attemptId = attemptId
+    this.registerShutdownHook = registerShutdownHook
+    this.shutdownHookPriority = shutdownHookPriority
+    logInfo(s"Starting Yarn extension services with app $appId and attemptId $attemptId")
+    super.start()
+  }
+
+  /**
+   * Loads a comma separated list of yarn services
+   */
+  override def serviceStart(): Unit = {
+    val sNames = sparkContext.getConf.getOption(YarnExtensionServices.SPARK_YARN_SERVICES)
```

Comment:
I prefer the slightly less noisy approach:

    services = sparkContext.getConf.getOption(YarnExtensionServices.SPARK_YARN_SERVICES)
      .map(_.split(","))
      .flatMap { sClass => ...
      }

You lose the "no services" debug message but it's not like it's useful.